boto/endpoints.py

```diff
@@ -0,0 +1,415 @@
+        # A mapping can be provided that handles the mapping between
+        # "service names" and endpoint prefixes.
+        'awslambda': 'lambda',
+        'cloudwatch': 'monitoring',
+        'ses': 'email',
+        'ec2containerservice': 'ecs',
+        'configservice': 'config',
+    }
+
+    def __init__(self, endpoint_data, boto_endpoint_data=None,
+                 service_rename_map=None):
+        """
+        :type endpoint_data: dict
+        :param endpoint_data: Regions and endpoints data in the same format
+            as is used by botocore / boto3.
+
+        :type boto_endpoint_data: dict
+        :param boto_endpoint_data: Regions and endpoints data in the legacy
+            boto2 format. This data takes precedence over any data found in
+            `endpoint_data`.
+
+        :type service_rename_map: dict
+        :param service_rename_map: A mapping of boto2 service name to
+            endpoint prefix.
+        """
+        super(BotoEndpointResolver, self).__init__(endpoint_data)
+        self._boto_endpoint_data = boto_endpoint_data
+        if service_rename_map is None:
+            service_rename_map = self.SERVICE_RENAMES
+        # Mapping of boto2 service name to endpoint prefix
+        self._endpoint_prefix_map = service_rename_map
+        # Mapping of endpoint prefix to boto2 service name
+        self._service_name_map = dict(
+            (v, k) for k, v in service_rename_map.items())
+
+    def get_available_endpoints(self, service_name, partition_name='aws',
+                                allow_non_regional=False):
+        endpoint_prefix = self._endpoint_prefix(service_name)
+        endpoints = super(BotoEndpointResolver, self).get_available_endpoints(
+            endpoint_prefix, partition_name, allow_non_regional)
+        boto_endpoints = self._get_available_boto_endpoints(service_name)
+        return list(set(endpoints + boto_endpoints))
+
+    def get_all_available_regions(self, service_name):
+        """Retrieve every region across partitions for a service."""
+        regions = set()
+        endpoint_prefix = self._endpoint_prefix(service_name)
+
+        # Get every region for every partition in the new endpoint format
+        for partition_name in self.get_available_partitions():
+            if self.is_global_service(service_name, partition_name):
+                # Global services are available in every region in the
+                # partition in which they are considered global.
+                partition = self._get_partition_data(partition_name)
+                regions.update(partition['regions'].keys())
+            else:
+                regions.update(
+                    super(BotoEndpointResolver, self).get_available_endpoints(
+                        endpoint_prefix, partition_name
+                    )
+                )
+
+        # boto2 endpoint format has no partitions, so load them all.
+        boto_endpoints = self._get_available_boto_endpoints(service_name)
```

Comment:
I don't think it makes sense to leak the concerns of resolving an endpoint to a builder. I could understand moving it to a different class, but then you'll have to re-implement the prioritization logic everywhere or create a third class that stitches the two together.

We won't be able to side-step the fact that we have to support both formats side-by-side. Currently the legacy format is only touched directly by functions that deal only with that format. The names of said functions specifically call out that they're working with that format. I agree with your other comment that it should be referenced in a more consistent way, and think that will help any confusion. The new format is only touched directly in cases where needed information is not provided by any existing functions. We could move that functionality up into the `EndpointResolver` to eliminate touching it directly at all.

Regarding the fallback question, I don't think the risk is really impacted either way. We still have to go through all of the new code to resolve all the regions. In fact, we have to go through *more* of it to resolve everything (since we need the static endpoint builder in addition to the endpoint resolver).

Furthermore, resolving every endpoint for every service is incredibly slow. For a known region it's ~4.5x slower than making a single call to `resolve_hostname` and for an unknown region it's a full order of magnitude slower.